src/caffe/util/im2col.cpp

```diff
@@ -6,44 +6,92 @@
+    int kernel_h, int kernel_w, int pad_h, int pad_w,
+    int stride_h, int stride_w, int dilation_h, int dilation_w,
+    Dtype *data_im) {
+  const int output_h = (input_h + 2 * pad_h -
+    (dilation_h * (kernel_h - 1) + 1)) / stride_h + 1;
+  const int output_w = (input_w + 2 * pad_w -
+    (dilation_w * (kernel_w - 1) + 1)) / stride_w + 1;
+  const int channel_size = input_h * input_w;
+  caffe_set(channels * input_h * input_w, Dtype(0), data_im);
+  for (; channels--; data_im += channel_size) {
+    for (int kernel_row = 0; kernel_row < kernel_h; kernel_row++) {
+      for (int kernel_col = 0; kernel_col < kernel_w; kernel_col++) {
+        int input_row = -pad_h + kernel_row * dilation_h;
+        for (int output_rows = output_h; output_rows; output_rows--) {
+          if (((unsigned) input_row) >= ((unsigned) input_h)) {
+            data_col += output_w;
+          } else {
+            int input_col = -pad_w + kernel_col * dilation_w;
+            for (int output_col = output_w; output_col; output_col--) {
+              if (((unsigned) input_col) < ((unsigned) input_w))
+                data_im[input_row * input_w + input_col] += *data_col;
+              data_col++;
+              input_col += stride_w;
+            }
+          }
+          input_row += stride_h;
+        }
       }
     }
   }
 }
 
 // Explicit instantiation
-template void im2col_cpu<float>(const float* data_im, const int channels,
-    const int height, const int width, const int kernel_h, const int kernel_w,
-    const int pad_h, const int pad_w, const int stride_h,
-    const int stride_w, const int dilation_h, const int dilation_w,
-    float* data_col);
-template void im2col_cpu<double>(const double* data_im, const int channels,
-    const int height, const int width, const int kernel_h, const int kernel_w,
-    const int pad_h, const int pad_w, const int stride_h,
-    const int stride_w, const int dilation_h, const int dilation_w,
-    double* data_col);
+template void im2col_cpu<float>(const float* data_im, int channels,
+    int input_h, int input_w, int kernel_h, int kernel_w,
+    int pad_h, int pad_w, int stride_h, int stride_w,
+    int dilation_h, int dilation_w, float* data_col);
+template void im2col_cpu<double>(const double* data_im, int channels,
+    int input_h, int input_w, int kernel_h, int kernel_w,
+    int pad_h, int pad_w, int stride_h, int stride_w,
+    int dilation_h, int dilation_w, double* data_col);
+template void col2im_cpu<float>(const float* data_col, int channels,
+    int input_h, int input_w, int kernel_h, int kernel_w,
+    int pad_h, int pad_w, int stride_h, int stride_w,
+    int dilation_h, int dilation_w, float* data_im);
+template void col2im_cpu<double>(const double* data_col, int channels,
+    int input_h, int input_w, int kernel_h, int kernel_w,
+    int pad_h, int pad_w, int stride_h, int stride_w,
+    int dilation_h, int dilation_w, double* data_im);
```

Comment:
(This patch would be easier to read if we didn't reorder the instantiations. It's okay as is, but in general let's factor out irrelevant changes.)